Introduce updatable user agent string quirk for storage access via WebPrivacy

sysrqb opened the pull request:
#### bceb6904160f321504dfccf7d2d5fe8f0a91b7bd
<pre>
Introduce updatable user agent string quirk for storage access via WebPrivacy
<a href="https://bugs.webkit.org/show_bug.cgi?id=265811">https://bugs.webkit.org/show_bug.cgi?id=265811</a>
<a href="https://rdar.apple.com/119148523">rdar://119148523</a>

Reviewed by Wenson Hsieh.

Currently we only have support for hard-coded user agent string quirks. This
patch introduces a new remotely updatable mechanism for quirking a request&apos;s
user agent string when it relates to quirking for storage access. On some sites
quirking the storage access prompt isn&apos;t sufficient, and those sites require
quirking the user agent string, as well.

* Source/WebCore/PAL/pal/cocoa/WebPrivacySoftLink.h:
* Source/WebCore/PAL/pal/cocoa/WebPrivacySoftLink.mm:
* Source/WebCore/PAL/pal/spi/cocoa/WebPrivacySPI.h:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::userAgent const):
* Source/WebCore/page/Quirks.cpp:
(WebCore::updatableStorageAccessUserAgentStringQuirks):
(WebCore::Quirks::updateStorageAccessUserAgentStringQuirks):
(WebCore::Quirks::storageAccessUserAgentStringQuirkForDomain):
* Source/WebCore/page/Quirks.h:
* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.h:
(WebKit::StorageAccessUserAgentStringQuirkObserver::create):
(WebKit::StorageAccessUserAgentStringQuirkObserver::invokeCallback):
(WebKit::StorageAccessUserAgentStringQuirkObserver::StorageAccessUserAgentStringQuirkObserver):
(WebKit::StorageAccessUserAgentStringQuirkController::cachedQuirks const):
* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
(-[WKWebPrivacyNotificationListener listenForStorageAccessUserAgentStringQuirkChanges:]):
(-[WKWebPrivacyNotificationListener didUpdate:]):
(WebKit::StorageAccessUserAgentStringQuirkController::shared):
(WebKit::StorageAccessUserAgentStringQuirkController::observeUpdates):
(WebKit::StorageAccessUserAgentStringQuirkController::setCachedQuirks):
(WebKit::StorageAccessUserAgentStringQuirkController::setCachedQuirksForTesting):
(WebKit::StorageAccessUserAgentStringQuirkController::updateQuirks):
* Source/WebKit/Shared/WebProcessCreationParameters.h:
* Source/WebKit/Shared/WebProcessCreationParameters.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _setUserAgentStringQuirkForTesting:withUserAgent:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::platformInitializeWebProcess):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::setUserAgentStringQuirkForTesting):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::initializeWebProcess):
(WebKit::WebProcess::updateStorageAccessUserAgentStringQuirks):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ResourceLoadStatistics.mm:
(TEST):

Canonical link: <a href="https://commits.webkit.org/271815@main">https://commits.webkit.org/271815@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2a25e4b095b707fa77eb827a9ddd583d68be4d17

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29723 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/8381 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/31022 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/32224 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26879 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/30342 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/10536 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5652 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/26892 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29999 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6995 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/25342 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5966 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/6118 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/26430 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33562 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/27091 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26854 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/32315 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/6055 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/4261 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/30097 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7799 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7057 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/6807 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/6588 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

charliewolfe commented on Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
```diff
@@ -729,6 +730,27 @@ bool ArgumentCoder<ResourceError>::decode(Decoder& decoder, ResourceError& resou
     return true;
 }
 
+void ArgumentCoder<UserAgentStringQuirk>::encode(Encoder& encoder, const UserAgentStringQuirk& userAgentStringQuirk)
```
I think this should go in WebCoreArgumentCoders.serialization.in

sysrqb replied:
Good call

charliewolfe commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -2937,7 +2938,10 @@ int FrameLoader::numPendingOrLoadingRequests(bool recurse) const
 {
     String userAgent;
-    if (RefPtr localFrame = dynamicDowncast<LocalFrame>(m_frame->mainFrame())) {
+    if (auto quirk = Quirks::storageAccessUserAgentStringQuirkForDomain(url); !quirk.isEmpty())
```
This looks like it will quirk the UA for any URL which matches a domain on our list, so nothing is really restricting this to be related to storage access. Is that okay?

sysrqb replied:
Good question. I think it should be okay because if we know that we need a UA quirk for a site to work, even after quirking the storage access prompt, and we decide that using the UA quirk is safe in that case, then using the UA quirk without the storage access prompt quirk should be safe, too. We could couple these more tightly, but I'll leave that to a follow up patch if we decide that's important.

charliewolfe commented on Source/WebCore/platform/network/UserAgentStringQuirk.h:
```diff
@@ -0,0 +1,35 @@
+namespace WebCore {
+
+struct UserAgentStringQuirk {
+    RegistrableDomain site;
```
Nit: I usually see RegistrableDomain variables named domain instead of site

sysrqb replied:
True, `domain` is vague, but I'll follow the current pattern.

charliewolfe commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.h:
```diff
@@ -87,6 +88,48 @@ class LinkDecorationFilteringController {
+    Function<void()> m_callback;
+};
+
+class StorageAccessUserAgentStringQuirkController {
```
We could probably reduce some code duplication between these classes and `LinkDecorationFilteringController`/`LinkDecorationFilteringDataObserver`, but I don't think it's too important.

whsieh replied:
Yeah, I think that would definitely be nice as well, but could be done as a follow up task.

We could probably use templated classes and/or helper functions to extract some of the common logic.

charliewolfe commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -2402,6 +2415,19 @@ void WebProcessProxy::systemBeep()
     PAL::systemBeep();
 }
 
+void WebProcessProxy::sendStorageAccessUserAgentStringQuirksData()
```
Should this be in WebPageProxy instead? We send the link decoration filtering data from there.

sysrqb replied:
We could. I'm using the WebProcess because these quirks aren't specific to any web page. I guess that's true for LDF data, too, though. I started by following how we do it for link decoration filtering, but it seemed more complicated with multiple checks in `decidePolicyForNavigationAction`. But I guess WebProcessProxy has its own complexity due to `shouldSendPendingMessage`.

cdumez commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -359,6 +360,30 @@ bool Quirks::shouldAvoidUsingIOS17UserAgentForFacebook() const
 
+HashMap<RegistrableDomain, UserAgentStringQuirk>& Quirks::updatableStorageAccessUserAgentStringQuirks()
+{
+    static NeverDestroyed<HashMap<RegistrableDomain, UserAgentStringQuirk>> map;
```
`MainThreadNeverDestroyed<>`

cdumez commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -359,6 +360,30 @@ bool Quirks::shouldAvoidUsingIOS17UserAgentForFacebook() const
+    auto& quirks = updatableStorageAccessUserAgentStringQuirks();
+    quirks.clear();
+    for (auto&& quirk : userAgentStringQuirks)
+        quirks.set(quirk.domain, quirk);
```
This should be a call to `add()`, not `set()` I think. Since you've just cleared the HashMap, there shouldn't be any need to overwrite. There is a possibility that there are several quirks for the same domain in the vector I guess. Not sure if that's possible but even if it is, it is unclear using the last matching quirk instead of the first matching one is better.

add() is more efficient.

cdumez commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -359,6 +360,30 @@ bool Quirks::shouldAvoidUsingIOS17UserAgentForFacebook() const
+{
+    const auto& quirks = updatableStorageAccessUserAgentStringQuirks();
+    RegistrableDomain domain { url };
+    auto quirk = quirks.find(domain);
```
Maybe we should call this `iterator` instead of `quirk`.

cdumez commented on Source/WebCore/page/Quirks.h:
```diff
@@ -98,6 +100,10 @@ class Quirks {
     WEBCORE_EXPORT bool shouldAvoidUsingIOS13ForGmail() const;
     WEBCORE_EXPORT bool shouldAvoidUsingIOS17UserAgentForFacebook() const;
 
+    WEBCORE_EXPORT static HashMap<RegistrableDomain, UserAgentStringQuirk>& updatableStorageAccessUserAgentStringQuirks();
```
Why is this public? Can't it be just a static function in the cpp?

cdumez commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.h:
```diff
@@ -87,6 +88,48 @@ class LinkDecorationFilteringController {
+private:
+    friend class StorageAccessUserAgentStringQuirkController;
+
+    StorageAccessUserAgentStringQuirkObserver(Function<void()>&& callback)
```
`explicit`.

cdumez commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.h:
```diff
@@ -87,6 +88,48 @@ class LinkDecorationFilteringController {
+    {
+    }
+
+    static Ref<StorageAccessUserAgentStringQuirkObserver> create(Function<void()>&& callback)
```
Why do we want this to be private? We usually have these public.

sysrqb replied:
I'm not sure, I was following the existing example. @charliewolfe @whsieh do you remember why these are private? (and below)

cdumez commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.h:
```diff
@@ -87,6 +88,48 @@ class LinkDecorationFilteringController {
+        return adoptRef(*new StorageAccessUserAgentStringQuirkObserver(WTFMove(callback)));
+    }
+
+    void invokeCallback() { m_callback(); }
```
Ditto, should this be public? Why using a "friend" here?

whsieh replied:
I think this is because we'd only expect `invokeCallback` to be invoked by the controller. Making it public seems okay, if we think that's cleaner than using `friend`.

cdumez commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
```diff
@@ -239,6 +249,71 @@ void requestLinkDecorationFilteringData(LinkFilteringRulesCallback&& callback)
 
+StorageAccessUserAgentStringQuirkController& StorageAccessUserAgentStringQuirkController::shared()
+{
+    static StorageAccessUserAgentStringQuirkController* sharedInstance = new StorageAccessUserAgentStringQuirkController;
```
Please use `MainThreadNeverDestroyed<>`.

cdumez commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
```diff
@@ -239,6 +249,71 @@ void requestLinkDecorationFilteringData(LinkFilteringRulesCallback&& callback)
+
+Ref<StorageAccessUserAgentStringQuirkObserver> StorageAccessUserAgentStringQuirkController::observeUpdates(Function<void()>&& callback)
+{
+    if (!m_notificationListener) {
```
We should assert we're on the main thread.

cdumez commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
```diff
@@ -239,6 +249,71 @@ void requestLinkDecorationFilteringData(LinkFilteringRulesCallback&& callback)
+            });
+        }];
+    }
+    auto observer = StorageAccessUserAgentStringQuirkObserver::create(WTFMove(callback));
```
`auto` -> `Ref`

cdumez commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
```diff
@@ -239,6 +249,71 @@ void requestLinkDecorationFilteringData(LinkFilteringRulesCallback&& callback)
+
+void StorageAccessUserAgentStringQuirkController::updateQuirks(CompletionHandler<void()>&& callback)
+{
+    if (!WebKit::canUseWebPrivacyFramework()) {
```
We should assert that we're on the main thread.

cdumez commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
```diff
@@ -239,6 +249,71 @@ void requestLinkDecorationFilteringData(LinkFilteringRulesCallback&& callback)
+    setCachedQuirks(WTFMove(quirks));
+}
+
+void StorageAccessUserAgentStringQuirkController::updateQuirks(CompletionHandler<void()>&& callback)
```
we should call it `completionHandler` instead of `callback` given the type.

cdumez commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
```diff
@@ -239,6 +249,71 @@ void requestLinkDecorationFilteringData(LinkFilteringRulesCallback&& callback)
+    if (lookupCallbacks->size() > 1)
+        return;
+
+    auto options = adoptNS([PAL::allocWPResourceRequestOptionsInstance() init]);
```
`auto` -> `RetainPtr`

cdumez commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
```diff
@@ -239,6 +249,71 @@ void requestLinkDecorationFilteringData(LinkFilteringRulesCallback&& callback)
+    [options setAfterUpdates:NO];
+
+    [[PAL::getWPResourcesClass() sharedInstance] requestStorageAccessUserAgentStringQuirksData:options.get() completionHandler:^(WPStorageAccessUserAgentStringQuirksData *data, NSError *error) {
+        Vector<WebCore::UserAgentStringQuirk> result;
```
Any reason we're using a `Vector<WebCore::UserAgentStringQuirk>` in this class? Why can't we store a `HashMap<RegistrableDomain, String /* user agent */>` here, since it seems to be what we need eventually? Why do we need the UserAgentStringQuirk struct at all?

cdumez commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
```diff
@@ -239,6 +249,71 @@ void requestLinkDecorationFilteringData(LinkFilteringRulesCallback&& callback)
+            setCachedQuirks(WTFMove(result));
+        }
+
+        for (auto& callback : std::exchange(lookupCallbacks.get(), { }))
```
completionHandler

cdumez commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -6804,3 +6804,8 @@ enum class WebCore::SourceBufferAppendMode : uint8_t {
 
 #endif
+
+struct WebCore::UserAgentStringQuirk {
```
I'm a little unclear why we need this struct at all.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -540,6 +548,9 @@ void WebProcessProxy::platformGetLaunchOptions(ProcessLauncher::LaunchOptions& l
 bool WebProcessProxy::shouldSendPendingMessage(const PendingMessage& message)
 {
+    // Send the quirk before handling any load requests.
+    sendStorageAccessUserAgentStringQuirksData();
```
Seems like an odd place to do this. Could you explain why calling this in didFinishLaunching() does not suffice? I'd expect didFinishLaunching() to be one of the first thing to get called.

Maybe you just need to call `sendStorageAccessUserAgentStringQuirksData()` inside `didFinishLaunching()` BEFORE the call to `AuxiliaryProcessProxy::didFinishLaunching()`, not after.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -1285,6 +1296,8 @@ void WebProcessProxy::didFinishLaunching(ProcessLauncher* launcher, IPC::Connect
 #endif
 #endif
 
+    sendStorageAccessUserAgentStringQuirksData();
```
Based on your change above, seems like you may want to call this before the call to the base class instead.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -2402,6 +2415,19 @@ void WebProcessProxy::systemBeep()
+void WebProcessProxy::sendStorageAccessUserAgentStringQuirksData()
+{
+#if ENABLE(ADVANCED_PRIVACY_PROTECTIONS)
+    if (!m_didSendStorageAccessUserAgentStringQuirks)
```
I think that if we call this in didFinishLaunching() once, like I suggested, then we probably don't need this flag. didFinishLaunching() can only get called once.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -2402,6 +2415,19 @@ void WebProcessProxy::systemBeep()
+        return;
+
+    // FIXME Filter by process's site when site isolation is enabled
+    send(Messages::WebProcess::SetStorageAccessUserAgentStringQuirks(StorageAccessUserAgentStringQuirkController::shared().cachedQuirks()), 0);
```
Actually, you know what, I am changing my mind. I don't think we need an IPC for this at all. Seems we should just add the quirks to the `WebProcessCreationParameters` struct.

whsieh commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -483,6 +483,11 @@ static void logProcessPoolState(const WebProcessPool& pool)
 #endif
+
+#if ENABLE(ADVANCED_PRIVACY_PROTECTIONS)
+    // FIXME Filter by process's site when site isolation is enabled
```
Nit - should be `// FIXME: Filter …`

whsieh commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -295,6 +296,18 @@ WebProcessPool::WebProcessPool(API::ProcessPoolConfiguration& configuration)
+        if (RefPtr protectedThis = weakThis.get()) {
+            if (StorageAccessUserAgentStringQuirkController::shared().cachedQuirks().isEmpty())
+                return;
+            // FIXME Filter by process's site when site isolation is enabled
```
Ditto.

whsieh commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ResourceLoadStatistics.mm:
```diff
@@ -1415,3 +1415,40 @@ - (void)webView:(WKWebView *)webView stopURLSchemeTask:(id <WKURLSchemeTask>)tas
+
+TEST(ResourceLoadStatistics, UserAgentStringForSite)
+{
+    auto configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
+    auto dataStore = adoptNS([[WKWebsiteDataStore alloc] _initWithConfiguration:adoptNS([[_WKWebsiteDataStoreConfiguration alloc] init]).get()]);
+    [configuration setWebsiteDataStore:dataStore.get()];
+
+    auto webView = adoptNS([[TestWKWebView alloc] initWithFrame:NSMakeRect(0, 0, 100, 100) configuration:configuration.get()]);
```
This seems like the default web view configuration; would it work if you just used `-initWithFrame:`?

```suggestion
    auto webView = adoptNS([[TestWKWebView alloc] initWithFrame:NSMakeRect(0, 0, 100, 100)]);
```

(And then replace `dataStore` below with `[webView configuration].dataStore`)

whsieh commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
```diff
@@ -330,7 +339,75 @@ void requestLinkDecorationFilteringData(LinkFilteringRulesCallback&& callback)
+    [[PAL::getWPResourcesClass() sharedInstance] requestStorageAccessUserAgentStringQuirksData:options.get() completionHandler:^(WPStorageAccessUserAgentStringQuirksData *data, NSError *error) {
+        HashMap<WebCore::RegistrableDomain, String> result;
+        if (error)
+            RELEASE_LOG_ERROR(ResourceLoadStatistics, "Failed to request user agent string quirks from WebPrivacy.");
```
Can we make this log message a tiny bit more specific?

```suggestion
            RELEASE_LOG_ERROR(ResourceLoadStatistics, "Failed to request storage access user agent string quirks from WebPrivacy.");
```

whsieh commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -359,6 +365,24 @@ bool Quirks::shouldAvoidUsingIOS17UserAgentForFacebook() const
+
+String Quirks::storageAccessUserAgentStringQuirkForDomain(const URL& url)
+{
+    const auto& quirks = updatableStorageAccessUserAgentStringQuirks();
```
It seems a bit odd that these are static, and not scoped to the document, like the other quirks (more concretely, this means that the quirk won't respect the value of `needsQuirks()`, which can be toggled by the user via Web Inspector UI).

Is this intentional?

sysrqb committed:
Introduce updatable user agent string quirk for storage access via WebPrivacy
https://bugs.webkit.org/show_bug.cgi?id=265811
rdar://119148523

Reviewed by Wenson Hsieh.

Currently we only have support for hard-coded user agent string quirks. This
patch introduces a new remotely updatable mechanism for quirking a request's
user agent string when it relates to quirking for storage access. On some sites
quirking the storage access prompt isn't sufficient, and those sites require
quirking the user agent string, as well.

* Source/WebCore/PAL/pal/cocoa/WebPrivacySoftLink.h:
* Source/WebCore/PAL/pal/cocoa/WebPrivacySoftLink.mm:
* Source/WebCore/PAL/pal/spi/cocoa/WebPrivacySPI.h:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::userAgent const):
* Source/WebCore/page/Quirks.cpp:
(WebCore::updatableStorageAccessUserAgentStringQuirks):
(WebCore::Quirks::updateStorageAccessUserAgentStringQuirks):
(WebCore::Quirks::storageAccessUserAgentStringQuirkForDomain):
* Source/WebCore/page/Quirks.h:
* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.h:
(WebKit::StorageAccessUserAgentStringQuirkObserver::create):
(WebKit::StorageAccessUserAgentStringQuirkObserver::invokeCallback):
(WebKit::StorageAccessUserAgentStringQuirkObserver::StorageAccessUserAgentStringQuirkObserver):
(WebKit::StorageAccessUserAgentStringQuirkController::cachedQuirks const):
* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
(-[WKWebPrivacyNotificationListener listenForStorageAccessUserAgentStringQuirkChanges:]):
(-[WKWebPrivacyNotificationListener didUpdate:]):
(WebKit::StorageAccessUserAgentStringQuirkController::shared):
(WebKit::StorageAccessUserAgentStringQuirkController::observeUpdates):
(WebKit::StorageAccessUserAgentStringQuirkController::setCachedQuirks):
(WebKit::StorageAccessUserAgentStringQuirkController::setCachedQuirksForTesting):
(WebKit::StorageAccessUserAgentStringQuirkController::updateQuirks):
* Source/WebKit/Shared/WebProcessCreationParameters.h:
* Source/WebKit/Shared/WebProcessCreationParameters.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _setUserAgentStringQuirkForTesting:withUserAgent:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::platformInitializeWebProcess):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::setUserAgentStringQuirkForTesting):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::initializeWebProcess):
(WebKit::WebProcess::updateStorageAccessUserAgentStringQuirks):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ResourceLoadStatistics.mm:
(TEST):

Canonical link: https://commits.webkit.org/271815@main
